File: soul-dashboard/server/session-cache.ts
/**
 * Session Cache - 세션별 이벤트 로컬 캐시
 *
 * 대시보드 서버가 Soul 서버에서 수신한 이벤트를 로컬에 캐시합니다.
 * 서버 재시작 시에도 캐시된 이벤트를 클라이언트에 즉시 전송할 수 있습니다.
 *
 * 파일 경로: {cacheDir}/{agentSessionId}.jsonl
 * 각 줄: {"id": <monotonic_int>, "event": <event_dict>}
 */

import { appendFile, readFile, mkdir, rm } from "fs/promises";
import { join, dirname } from "path";

export interface SessionCacheOptions {
  /** 캐시 파일 기본 디렉토리 */
  cacheDir: string;
}

export interface CachedEvent {
  id: number;
  event: Record<string, unknown>;
}

/**
 * 세션별 이벤트 JSONL 캐시.
 *
 * Soul 서버에서 수신한 이벤트를 로컬에 캐시하여
 * 서버 재시작 시에도 빠르게 클라이언트에 전송할 수 있습니다.
 */
export class SessionCache {
  private readonly cacheDir: string;

  constructor(options: SessionCacheOptions) {
    this.cacheDir = options.cacheDir;
  }

  /**
   * 이벤트를 세션 캐시에 추가합니다.
   *
   * @param agentSessionId - 세션 식별자
   * @param eventId - 이벤트 ID (단조증가)
   * @param event - 이벤트 데이터
   */
  async appendEvent(
    agentSessionId: string,
    eventId: number,
    event: Record<string, unknown>,
  ): Promise<void> {
    const filePath = this.sessionPath(agentSessionId);

    // 디렉토리 생성 (없으면)
    await mkdir(dirname(filePath), { recursive: true });

    // JSONL 형식으로 추가
    const record: CachedEvent = { id: eventId, event };
    const line = JSON.stringify(record) + "\n";

    await appendFile(filePath, line, "utf-8");
  }

  /**
   * 세션의 캐시된 이벤트를 읽습니다.
   *
   * @param agentSessionId - 세션 식별자
   * @param afterId - 이 ID 이후의 이벤트만 반환 (옵션)
   * @returns 이벤트 배열
   */
  async readEvents(
    agentSessionId: string,
    afterId?: number,
  ): Promise<CachedEvent[]> {
    const filePath = this.sessionPath(agentSessionId);

    let content: string;
    try {
      content = await readFile(filePath, "utf-8");
    } catch {
      // 파일이 없으면 빈 배열
      return [];
    }

    const events = this.parseJsonl(content);

    if (afterId !== undefined) {
      return events.filter((ev) => ev.id > afterId);
    }

    return events;
  }

  /**
   * 세션의 마지막 이벤트 ID를 반환합니다.
   *
   * @param agentSessionId - 세션 식별자
   * @returns 마지막 이벤트 ID (없으면 0)
   */
  async getLastEventId(agentSessionId: string): Promise<number> {
    const events = await this.readEvents(agentSessionId);

    if (events.length === 0) {
      return 0;
    }

    return events[events.length - 1].id;
  }

  /**
   * 세션 캐시를 삭제합니다.
   *
   * @param agentSessionId - 세션 식별자
   */
  async deleteSession(agentSessionId: string): Promise<void> {
    const filePath = this.sessionPath(agentSessionId);

    try {
      await rm(filePath, { force: true });
    } catch {
      // 파일이 없어도 오류 없이 처리
    }
  }

  /**
   * 세션 캐시 파일 경로를 반환합니다.
   */
  sessionPath(agentSessionId: string): string {
    const safeId = this.sanitize(agentSessionId);
    return join(this.cacheDir, `${safeId}.jsonl`);
  }

  /**
   * 경로 탈출 방지를 위한 문자열 sanitize.
   */
  private sanitize(value: string): string {
    return value.replace(/[^\w.\-]/g, "_");
  }

  /**
   * JSONL 문자열을 파싱합니다.
   */
  private parseJsonl(content: string): CachedEvent[] {
    const records: CachedEvent[] = [];
    const lines = content.split("\n");

    for (const line of lines) {
      const trimmed = line.trim();
      if (!trimmed) continue;

      try {
        const record = JSON.parse(trimmed) as CachedEvent;
        if (typeof record.id === "number" && record.event) {
          records.push(record);
        }
      } catch {
        console.warn("[SessionCache] Skipping corrupted JSONL line");
      }
    }

    return records;
  }
}

File: soul-server/src/soul_server/dashboard/session_cache.py
"""
SessionCache - 세션별 이벤트 로컬 캐시

soul-server가 수신한 이벤트를 로컬에 JSONL 형식으로 캐시합니다.
서버 재시작 시에도 캐시된 이벤트를 클라이언트에 즉시 전송할 수 있습니다.

파일 경로: {cache_dir}/{safe_session_id}.jsonl
각 줄: {"id": <monotonic_int>, "event": <event_dict>}
"""

import asyncio
import json
import pathlib
import re
from typing import Optional


class SessionCache:
    """세션별 이벤트 JSONL 캐시."""

    def __init__(self, cache_dir: str):
        self._dir = pathlib.Path(cache_dir)
        self._dir.mkdir(parents=True, exist_ok=True)

    def _path(self, session_id: str) -> pathlib.Path:
        safe = re.sub(r"[^\w.\-]", "_", session_id)
        return self._dir / f"{safe}.jsonl"

    async def append_event(self, session_id: str, event_id: int, event: dict) -> None:
        """이벤트를 세션 캐시에 추가합니다.

        동기 파일 I/O를 asyncio.to_thread로 래핑하여 이벤트 루프 블로킹을 방지합니다.
        """
        line = json.dumps({"id": event_id, "event": event}, ensure_ascii=False)
        path = self._path(session_id)

        def _write():
            with path.open("a", encoding="utf-8") as f:
                f.write(line + "\n")

        await asyncio.to_thread(_write)

    async def read_events(
        self, session_id: str, after_id: Optional[int] = None
    ) -> list:
        """세션의 캐시된 이벤트를 읽습니다.

        동기 파일 I/O를 asyncio.to_thread로 래핑하여 이벤트 루프 블로킹을 방지합니다.

        Args:
            session_id: 세션 식별자
            after_id: 이 ID 이후의 이벤트만 반환 (None이면 전체)

        Returns:
            {"id": int, "event": dict} 항목 리스트
        """
        path = self._path(session_id)
        if not path.exists():
            return []

        def _read():
            results = []
            for line in path.read_text(encoding="utf-8").splitlines():
                if not line.strip():
                    continue
                try:
                    item = json.loads(line)
                except json.JSONDecodeError:
                    continue
                if after_id is None or item["id"] > after_id:
                    results.append(item)
            return results

        return await asyncio.to_thread(_read)

    async def get_last_event_id(self, session_id: str) -> Optional[int]:
        """세션의 마지막 이벤트 ID를 반환합니다."""
        events = await self.read_events(session_id)
        return events[-1]["id"] if events else None

    async def delete_session(self, session_id: str) -> None:
        """세션 캐시를 삭제합니다."""
        path = self._path(session_id)
        if path.exists():
            path.unlink()
